### ISSUE TITLE
Feasabilty of enabling Trim on WinForms by manually providing the roots file?

### ISSUE THREAD

**Issue by jhudsoncedaron** [dotnet/winforms#7149]
### .NET version

.NET Runtime 6.0

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

N/A

### Issue description

Discussion tag please

This documentation describes how to provide a list of assemblies to preserve: https://github.com/dotnet/linker/blob/main/docs/data-formats.md

We have a large .NET Runtime application that is mostly ASP.NET Core, but also some Console, and a single small WinForms application whose sole purpose is to edit the configuration files. The team is really excited about trimmed single files and wants to publish the whole application to single file depolyment if possible. I've already solved the problem of how to amalgamate a bunch of applications to a single build for single file deployment.

What's the feasibility of manually providing the roots for COM interop endpoints so that ILLink sees them as rooted and doesn't remove them? The winforms application is so small we can walk all events on all controls in less than an hour and know full well whether it works or not.

All controls used in no particular order:
* Button
* ListView
* TextBox
* Label
* Panel
* TabControl
* DataGridView
* ComboBox
* TextBox
* CheckBox
* RichTextBox (mistake: a TextBox should have been used and I'll gladly switch it if it saves trouble)

Originally we shipped the Winforms binary as a .NET Framework binary; however we can no longer do so as this caused two issues:
* It is not possibly to load .NET Runtime assemblies from .NET Framework anymore, even to reflect on them rather than run code in them.
* The application did not work on Windows Server Core. We have since gone out of our way to make the config tool run on Server Core with the .NET Core/Runtime version of WinForms.

### Steps to reproduce

N/A

**Comment by RussKie** [comment]
Windows Forms SDK is not currently trim-friendly due to an extensive dependency on COM and the use of reflection. The team is aware of the customer ask in this area, however this work is not currently prioritised. We track related requests under https://github.com/dotnet/winforms/issues?q=label%3A%22area%3A+ILLinker%2FAOT%22
@kant2002 is currently porting COM definitions to ComWrappers, and when this work is complete, we may be in a position to start considering trimming annotations. That said, we welcome the community help and contributions in this area.

**Comment by kant2002** [comment]
@jhudsoncedaron I think you looking for https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options#root-descriptors

For starting place on controls which problematic for trimming, you can look at [my work](https://github.com/kant2002/RdXmlLibrary/blob/main/System.Windows.Forms.rd.xml) on NativeAOT. That's all problematic places which I found during my testing. It's likely that's not fully complete list, but that's good starting point. If you cover all tests you would be fine. Please note that format of RD.xml and `TrimmerRootDescriptor` are different, but they are close enough so you get a gist.

Regarding trimming annotations. External contributions are not blocked by ComWrappers work in any way right now, and I think process relatively easy, so community will helps. 

Not sure if you are able to make trimming for WinForms, but that's how I enable it right now.
```xml
<PropertyGroup>
  <_SuppressWinFormsTrimError>true</_SuppressWinFormsTrimError>
</PropertyGroup>
  
<ItemGroup>
  <TrimmableAssembly Include="System.Windows.Forms" />
  <TrimmableAssembly Include="System.Windows.Forms.Primitives" />
</ItemGroup>

<Target Name="ConfigureTrimming" BeforeTargets="PrepareForILLink">
  <ItemGroup>
    <ManagedAssemblyToLink Condition="'%(Filename)' == 'System.Windows.Forms'">
      <IsTrimmable>true</IsTrimmable>
    </ManagedAssemblyToLink>
    <ManagedAssemblyToLink Condition="'%(Filename)' == 'System.Windows.Forms.Primitives'">
      <IsTrimmable>true</IsTrimmable>
    </ManagedAssemblyToLink>
  </ItemGroup>
</Target>
```

**Comment by jhudsoncedaron** [comment]
@kant2002 : Deferred due to overload. Sorry. :(

**Comment by kant2002** [comment]
That’s unfortunate. Hopefully that’s not result in your overwork. And if you and your team still manage to run experiment I would still be interested to hear about lessons if this issue would still be with us

**Comment by TheJoeFin** [comment]
I have a very simple WinForms app which I have packaged into an MSIX for the Microsoft Store. I am experimenting with trying to trim the assembly and get it much smaller. Currently it builds and runs locally, but I have some issues when installing from the MSIX. If Anyone is interested in seeing how I am trying to get this working the app and package repos are here:

https://github.com/TheJoeFin/Caffeinated/tree/trimming-enabled
https://github.com/TheJoeFin/Caffeinated_Package/tree/trimming-enabled
